File: backend/chat/chart_generator.py
import json
from typing import Dict, Any, List

class ChartGenerator:
    """Genera configuraciones de gráficos basadas en resultados de queries"""
    
    @staticmethod
    def detect_chart_type(data: List[Dict], columns: List[str]) -> Dict[str, Any]:
        """Detecta el tipo de gráfico más apropiado"""
        if not data or len(data) == 0:
            return {'type': 'text', 'message': 'Sin datos para visualizar'}
        
        # Contar columnas numéricas
        numeric_cols = []
        categorical_cols = []
        
        for col in columns:
            if col not in data[0]:
                continue
            
            val = data[0][col]
            if isinstance(val, (int, float)):
                numeric_cols.append(col)
            else:
                categorical_cols.append(col)
        
        # Lógica de detección
        if len(categorical_cols) >= 1 and len(numeric_cols) >= 1:
            # Categoría + número = Bar chart
            return ChartGenerator._create_bar_chart(data, categorical_cols[0], numeric_cols[0])
        
        elif len(numeric_cols) >= 2:
            # Dos números = Scatter plot
            return ChartGenerator._create_scatter_chart(data, numeric_cols[0], numeric_cols[1])
        
        elif len(numeric_cols) == 1:
            # Un número = Distribución
            return ChartGenerator._create_distribution_chart(data, numeric_cols[0])
        
        else:
            return {'type': 'text', 'message': 'No se puede generar gráfico con estos datos'}
    
    @staticmethod
    def _create_bar_chart(data: List[Dict], x_col: str, y_col: str) -> Dict[str, Any]:
        """Crea configuración de bar chart"""
        chart_data = []
        for row in data[:20]:  # Limita a 20 filas
            chart_data.append({
                'name': str(row.get(x_col, 'N/A'))[:20],
                'value': float(row.get(y_col, 0)) if isinstance(row.get(y_col), (int, float)) else 0
            })
        
        return {
            'type': 'bar',
            'title': f'{y_col} por {x_col}',
            'data': chart_data,
            'x_axis': x_col,
            'y_axis': y_col,
        }
    
    @staticmethod
    def _create_scatter_chart(data: List[Dict], x_col: str, y_col: str) -> Dict[str, Any]:
        """Crea configuración de scatter plot"""
        chart_data = []
        for row in data[:50]:
            x_val = row.get(x_col)
            y_val = row.get(y_col)
            if isinstance(x_val, (int, float)) and isinstance(y_val, (int, float)):
                chart_data.append({
                    'x': float(x_val),
                    'y': float(y_val)
                })
        
        return {
            'type': 'scatter',
            'title': f'{y_col} vs {x_col}',
            'data': chart_data,
            'x_axis': x_col,
            'y_axis': y_col,
        }
    
    @staticmethod
    def _create_single_value_chart(data: List[Dict], col: str) -> Dict[str, Any]:
        """Crea un indicador visual para un solo valor numérico"""
        val = float(list(data[0].values())[0])
        return {
            'type': 'bar',
            'title': col.replace('_', ' ').title(),
            'data': [{'name': col.replace('_', ' ').title(), 'value': val}],
            'x_axis': 'Métrica',
            'y_axis': 'Valor',
        }

    @staticmethod
    def _create_distribution_chart(data: List[Dict], col: str) -> Dict[str, Any]:
        """Crea gráfico de distribución (histograma)"""
        values = []
        for row in data:
            val = row.get(col)
            if isinstance(val, (int, float)):
                values.append(float(val))
        
        if not values:
            return {'type': 'text', 'message': f'No hay valores numéricos en {col}'}
        
        if len(values) == 1:
            return ChartGenerator._create_single_value_chart(data, col)
        
        # Crear bins
        min_val = min(values)
        max_val = max(values)
        bin_count = min(10, len(set(values)))
        bin_size = (max_val - min_val) / bin_count if max_val != min_val else 1
        
        bins = {}
        for val in values:
            bin_idx = int((val - min_val) / bin_size) if bin_size > 0 else 0
            bin_idx = min(bin_idx, bin_count - 1)
            bin_label = f'{min_val + bin_idx * bin_size:.1f}'
            bins[bin_label] = bins.get(bin_label, 0) + 1
        
        chart_data = [
            {'range': label, 'count': count}
            for label, count in sorted(bins.items())
        ]
        
        return {
            'type': 'histogram',
            'title': f'Distribución de {col}',
            'data': chart_data,
            'column': col,
        }

File: frontend/src/engine/chartGenerator.js
// Elige un gráfico para el resultado de una consulta.
// Formas que entiende ChatMessages.jsx: bar/pie -> [{name, value}], scatter -> [{x, y}],
// histogram -> [{range, count}], text -> sin gráfico.

const isNum = (v) => typeof v === 'number'
const label = (v) => String(v ?? 'N/A').slice(0, 20)

// `listing`: la consulta es un SELECT * (filas crudas, no una agregación)
export function detectChartType(data, columns, { listing = false } = {}) {
  if (!data || data.length === 0) return { type: 'text', message: 'Sin datos para visualizar' }
  if (listing) return { type: 'text', message: 'Listado de registros' }

  const numericCols = []
  const categoricalCols = []
  for (const col of columns) {
    const values = data.map((row) => row[col]).filter((v) => v !== null && v !== undefined)
    if (values.length === 0) continue
    if (values.every(isNum)) numericCols.push(col)
    else categoricalCols.push(col)
  }

  // Un listado de filas (SELECT * ...) no es una agregación: no hay nada útil que graficar
  if (columns.length > 3 && data.length > 1) return { type: 'text', message: 'Listado de registros' }

  if (categoricalCols.length >= 1 && numericCols.length >= 1) {
    if (data.length === 1) return { type: 'text', message: 'Resultado único' }
    return data.length <= 8
      ? pieChart(data, categoricalCols[0], numericCols[0])
      : barChart(data, categoricalCols[0], numericCols[0])
  }
  if (categoricalCols.length >= 1) return pieChart(data, categoricalCols[0], null)

  if (numericCols.length >= 2) {
    // Una sola fila con varias métricas (promedios de varias columnas): barras por métrica
    if (data.length === 1) return metricsChart(data[0], numericCols)
    return scatterChart(data, numericCols[0], numericCols[1])
  }
  if (numericCols.length === 1) return distributionChart(data, numericCols[0])

  return { type: 'text', message: 'No se puede generar gráfico con estos datos' }
}

function pieChart(data, catCol, numCol) {
  return {
    type: 'pie',
    title: numCol ? `${numCol} por ${catCol}` : `Distribución de ${catCol}`,
    data: data.slice(0, 12).map((row) => ({
      name: label(row[catCol]),
      value: numCol && isNum(row[numCol]) ? row[numCol] : 1,
    })),
  }
}

function barChart(data, xCol, yCol) {
  return {
    type: 'bar',
    title: `${yCol} por ${xCol}`,
    data: data.slice(0, 20).map((row) => ({
      name: label(row[xCol]),
      value: isNum(row[yCol]) ? row[yCol] : 0,
    })),
    x_axis: xCol,
    y_axis: yCol,
  }
}

function metricsChart(row, numericCols) {
  return {
    type: 'bar',
    title: 'Métricas',
    data: numericCols.slice(0, 20).map((col) => ({ name: label(col), value: row[col] })),
    x_axis: 'Métrica',
    y_axis: 'Valor',
  }
}

function scatterChart(data, xCol, yCol) {
  return {
    type: 'scatter',
    title: `${yCol} vs ${xCol}`,
    data: data
      .slice(0, 50)
      .filter((row) => isNum(row[xCol]) && isNum(row[yCol]))
      .map((row) => ({ x: row[xCol], y: row[yCol] })),
    x_axis: xCol,
    y_axis: yCol,
  }
}

function distributionChart(data, col) {
  const values = data.map((row) => row[col]).filter(isNum)
  if (values.length === 0) return { type: 'text', message: `No hay valores numéricos en ${col}` }
  if (values.length === 1) {
    return {
      type: 'bar',
      title: col.replace(/_/g, ' '),
      data: [{ name: col.replace(/_/g, ' '), value: values[0] }],
      x_axis: 'Métrica',
      y_axis: 'Valor',
    }
  }

  const min = Math.min(...values)
  const max = Math.max(...values)
  const binCount = Math.min(10, new Set(values).size)
  const binSize = max !== min ? (max - min) / binCount : 1
  const counts = new Array(binCount).fill(0)
  for (const v of values) {
    counts[Math.min(Math.floor((v - min) / binSize), binCount - 1)]++
  }
  const decimals = binSize >= 1 ? 1 : binSize >= 0.1 ? 2 : 4

  return {
    type: 'histogram',
    title: `Distribución de ${col}`,
    data: counts.map((count, i) => ({ range: (min + i * binSize).toFixed(decimals), count })),
    column: col,
  }
}
